wagtailmodelchoosers/rich_text.py

```diff
@@ -0,0 +1,122 @@
+    """
+    Lower and replace non-alphanumeric characters with underscores.
+    """
+
+    return re.sub(r'[^\w]', '_', name.lower())
+
+
+def get_chooser_generic_to_database_format(feature_name):
+    entity_name = normalize_entity_name(feature_name)
+    attrname = 'data-{}'.format(entity_name.replace('_', '-'))
+
+    def entity_decorator(props):
+        return DOM.create_element(
+            'span',
+            {attrname: props['value']},
+            props['children'],
+        )
+    entity_decorator.__name__ = '{}_entity_decorator'.format(feature_name)
+
+    return entity_decorator
+
+
+def get_chooser_generic_from_database_format(feature_name):
+    entity_name = normalize_entity_name(feature_name)
+    attrname = 'data-{}'.format(entity_name.replace('_', '-'))
+    klassname = '{}EntityElementHandler'.format(entity_name.title().replace('_', ''))
+
+    class BaseEntityElementHandler(InlineEntityElementHandler):
+        mutability = 'IMMUTABLE'
+
+        def get_attribute_data(self, attrs):
+            return {
+                'value': attrs[attrname],
+            }
+
+    return type(klassname, (BaseEntityElementHandler,), {})
+
+
+def get_chooser_feature(
+    chooser,
+    feature_name,
+    feature_type,
+    icon=None,
+    label=None,
+    description=None,
+    from_database_format=None,
+    to_database_format=None,
+    js_source='window.wagtailModelChoosers.ModelSource',
+    js_decorator='window.wagtailModelChoosers.GenericModelDecorator',
+):
+
+    # Base control options
+    chooser_options = get_chooser_options('simple_model')
+    control = dict(chooser_options, **{
+        'type': feature_type,
+    })
+
+    # Add overwrites if any
+    if icon:
+        control['icon'] = icon
+    if label:
+        control['label'] = label
+    if description:
+        control['description'] = description
+
+    # Get converters
+    from_database_format = from_database_format or get_chooser_generic_from_database_format(feature_name)
+    to_database_format = to_database_format or get_chooser_generic_to_database_format(feature_name)
+
+    # Create register feature function
+    def register_feature(features):
+        # Register the feature
+        features.register_editor_plugin(
+            'draftail',
+            feature_name,
+            draftail_features.EntityFeature(control),
+        )
+
+        # Register the converter rules
+        features.register_converter_rule(
+            'contentstate',
+            feature_name,
+            {
+                'from_database_format': {'span[data-stock]': from_database_format(feature_type)},  # TODO: Replace stock
+                'to_database_format': {'entity_decorators': {feature_type: to_database_format}},
+            }
+        )
+
+    # Create register plugin scripts
+    js_dependencies = (
+        'wagtailadmin/js/draftail.js',
+        'wagtailmodelchoosers/wagtailmodelchoosers.js',
+        'wagtailmodelchoosers/polyfills.js',
```

Comment:
It feels a bit funny to see the polyfills loaded after the rest of the code. They mostly target `window.fetch` – I think this could easily be replaced by jQuery's XMLHttpRequest wrappers, which would have the advantage of supporting cancelation (since this is used for autocomplete fields, it's quite valuable).